scipy/datasets/_utils.py

```diff
@@ -0,0 +1,62 @@
+import os
+import shutil
+from ._registry import registry
+
+try:
+    import appdirs
+except ImportError:
+    appdirs = None
+
+
+def _clear_cache(datasets, cache_dir=None, data_registry=None):
+    if data_registry is None:
+        # Use SciPy Datasets registry map
+        data_registry = registry
+    if cache_dir is None:
+        # Use default cache_dir path
+        if appdirs is None:
+            # appdirs is pooch dependency
+            raise ImportError("Missing optional dependency 'pooch' required "
+                              "for scipy.datasets module. Please use pip or "
+                              "conda to install 'pooch'.")
+        cache_dir = appdirs.user_cache_dir("scipy-data")
+
+    if not os.path.exists(cache_dir):
+        print(f"Cache Directory {cache_dir} doesn't exist. Nothing to clear.")
+        return
+
+    if datasets is None:
+        print(f"Cleaning the cache directory {cache_dir}!")
+        shutil.rmtree(cache_dir)
+    else:
+        for dataset_name in datasets:
+            if dataset_name not in data_registry:
+                raise ValueError(f"Dataset {dataset_name} doesn't exist. "
+                                 "Please check if the passed dataset "
+                                 "is a subset of the following: "
+                                 f"{list(data_registry.keys())}")
+
+            dataset_path = os.path.join(cache_dir, dataset_name)
+            if os.path.exists(dataset_path):
+                print(f"Cleaning the {dataset_name} file!")
+                os.remove(dataset_path)
+            else:
+                print(f"Path {dataset_path} doesn't exist. Nothing to clear.")
+
+
+def clear_cache(datasets=None):
+    """
+    Cleans the scipy datasets cache directory.
+
+    If a list/tuple of dataset strings is provided, then it
+    removes all the datasets in the list/tuple.
+
+    By default, it removes all the cached data files.
+
+    Parameters
+    ----------
+    datasets : list/tuple of str or None
+        A list/tuple of datasets to be removed from cache.
```

Comment:
This seems to be doing the wrong thing:
```python
ValueError: Dataset a doesn't exist. Please check if the passed dataset is a subset of
the following: ['ascent.dat', 'ecg.dat', 'face.dat']
```
Those file names are private; they may change over time for some datasets, and there may be multiple data files for a single dataset. 

I suggest taking the function itself as input here, or a tuple of functions. So:
```
>>> datasets.clear_cache(datasets.ascent)  # removes any data files associated with `ascent`
```

Reply:
The string form is broken anyway, it looks like a string is treated as a sequence instead:
```
>>> d.clear_cache('ascent.dat')
ValueError: Dataset a doesn't exist. Please check if the passed dataset is a subset of the following: ['ascent.dat', 'ecg.dat', 'face.dat']

>>> d.clear_cache(('ascent.dat',))  # tuple format does work
Path /home/rgommers/.cache/scipy-data/ascent.dat doesn't exist. Nothing to clear.

Reply:
> I suggest taking the function itself as input here, or a tuple of functions:

The functions internally hardcodes the `filename` for the datasets associated, and the `filename` is never returned from the datasets function itself. See these lines https://github.com/scipy/scipy/blob/276f9e0e2319fca37adc7e4689a3316808c7fca3/scipy/datasets/_fetchers.py#L71
https://github.com/scipy/scipy/blob/276f9e0e2319fca37adc7e4689a3316808c7fca3/scipy/datasets/_fetchers.py#L169
https://github.com/scipy/scipy/blob/276f9e0e2319fca37adc7e4689a3316808c7fca3/scipy/datasets/_fetchers.py#L211

So I don't see a way to retrieve dataset files from the function name  unless we manually create and maintain a map somewhere (maybe in https://github.com/scipy/scipy/blob/main/scipy/datasets/_registry.py which is anyway updated everytime a new dataset is added). If we do that, then we can look at the map and remove all the files corresponding to the function name. Also then we will not hardcode the above lines in dataset methods, instead we can check the map to get the filename when fetching it. So essentially, there will be only one place to maintain the dataset function names and corresponding filenames map.

> The string form is broken anyway, it looks like a string is treated as a sequence instead:

Docstring specifically mentions the use of tuples and list only. I should have made an assert check on the passed argument for a better error in that case.
Other option was to convert the string to a list internally and carry on with the same execution, which i had before but i decided to drop that to simplify the signature. I can add that back if we continue with the string form.

In any case I agree that passing the function itself is a better API than the user having to think about what the filename will be.

Reply:
> Docstring specifically mentions the use of tuples and list only

Ah, I misread the "of str" part - late night reviewing.



> So essentially, there will be only one place to maintain the dataset function names and corresponding filenames map.

That seems like a good idea. A tweak like this should do it:
```python
registry = {
    "ascent": {"ascent.dat": "e8a84939484463ab8051aedc5b40aa262ab33a91d6458a6cd13c6a1cad5a023d"},
```

Reply:
Altering the registry is not something we'd want since `pooch` takes this dictionary in the specific format that we have right now.
```
registry = {
    "<filename>": "<file_hash>",
}
```

I ended up creating a new dict `method_files_map` to handle this. Right now we don't have any dataset method which fetches multiple files, but the updated `clear_cache` method should handle this if we do add such a method in the future. Though this ended up being a lot more complicated than I initially would have liked.

Btw, I'd expect if anyone uses this function, they will usually prefer calling just the default setting, clearing all the cache, and that doesn't require any arguments.

> Also then we will not hardcode the above lines in dataset methods, instead, we can check the map to get the filename when fetching it.

I wanted to avoid hardcoding the filenames inside the actual dataset methods but it turns out there is no direct way to get the function name string for accessing `method_files_map` inside the function itself without using `inspect`. If we go that route with `inspect`, it just makes it more complicated for no good reason, so I kept the hardcoded filenames.

I've also added a small example in the docstring to avoid confusion for the user.